Technical/Logging/The-Tazama-Logging-Framework.md

```diff
@@ -1 +1,405 @@
+```
+> [!CAUTION]
+> Ensure the sidecar AND [lumberjack] start up before the processor so that logs are not missed.
+
+##### Without Event-Sidecar and Lumberjack
+In your processor, add the [frms-coe-lib] as a dependency:
+
+```sh
+npm i @frmscoe/frms-coe-lib
+```
+
+Configure your environment:
+| Variable Name         | Purpose                               | Example             |
+|-----------------------|---------------------------------------|---------------------|
+| ELASTIC_SEARCH_VERSION| Specifies the Elasticsearch version   | `8.11`              |
+| ELASTIC_HOST          | Specifies the Elasticsearch host       | `http://localhost:9200` |
+| ELASTIC_USERNAME      | Username for Elasticsearch authentication | `elastic_user`   |
+| ELASTIC_PASSWORD      | Password for Elasticsearch authentication | `secretpassword` |
+
+In your processor, create an instance of a `LoggerService`:
+
+```ts
+import { LoggerService } from '@frmscoe/frms-coe-lib';
+
+const logger = new LoggerService();
+```
+#### Usage
+
+You may specify a `LOGSTASH_LEVEL` environment variable in your processor to control your logs. The logging system supports multiple levels of severity, listed in ascending order of importance: `trace`, `debug`, `info`, `warn`, `error`, and `fatal`. When configuring the logging level, every log event with that level or a higher severity will be logged. For example, if you set the logging level to `info`, all log events at `info`, `warn`, `error`, and `fatal` levels will be logged. This ensures that important information and higher severity events are captured while filtering out less critical details.
+
+After initialising your logger, you may use the methods available to the `LoggerService` instance.
+
+##### Logger Service 
+
+
+###### Function Signature:
+
+The custom logger function has the following signature for `trace`, `debug`, `log` and `warn` log levels:
+
+```typescript
+trace(message: string, serviceOperation?: string, id?: string, callback?: LogCallback): void
+debug(message: string, serviceOperation?: string, id?: string, callback?: LogCallback): void
+log(message: string, serviceOperation?: string, id?: string, callback?: LogCallback): void
+```
+
+- **`message` (required)**: A string that represents the log message to be recorded.
+
+- **`serviceOperation` (optional)**: Specifies the name or identifier of the service operation related to the log message. This parameter provides context about the operation being logged.
+
+- **`id` (optional)**: An optional identifier associated with the log message. Use this parameter to correlate logs related to a specific transaction or request.
+
+- **`callback` (optional)**: A callback function that can be provided to handle asynchronous logging operations or to receive notifications after logging. This parameter is optional and may not be used in all logging scenarios.
+
+Additional Levels:
+`error` and `fatal` levels are also supported but they have a custom signature:
```

Comment:
As above, I see why this is a separate function (additional parameters) and also, curiously, here we are combining `error` and `fatal` into a single function?

Reply:
I will provide more clarity on this. Basically, we are **not** combining them. They are separate functions but they have the same signature (which is different from the signature of the log levels lower than `error`):


trace,debug,log,warn => Function Signature A
error,fatal => Function Signature B